### PR TITLE
feat(ingestion): azure-ad stateful ingestion

### PR DESCRIPTION
**Pull request by mohdsiddique**

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py [hunk 1]
```diff
@@ -529,7 +568,9 @@ def _map_azure_ad_group(self, azure_ad_group):
             self.config.azure_ad_response_to_groupname_regex,
         )
         if not self.config.groups_pattern.allowed(group_name):
-            self.report.report_filtered(f"{corp_group_urn}")
+            # lintFix giving error: "StatefulIngestionReport" has no attribute "report_filtered"
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py:571]: Add a type annotation for the report type in this class. At the top of the class, add `report: AzureSourceReport` or similar

Do not use type ignore here

File: metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py [hunk 2]
```diff
@@ -575,7 +616,9 @@ def _map_azure_ad_users(self, azure_ad_users):
             if error_str is not None:
                 continue
             if not self.config.users_pattern.allowed(corp_user_urn):
-                self.report.report_filtered(f"{corp_user_urn}.*")
+                # lintFix giving error: "StatefulIngestionReport" has no attribute "report_filtered"
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py:619]: same as above

File: metadata-ingestion/tests/integration/azure_ad/test_azure_ad.py [hunk 3]
```diff
@@ -308,6 +315,121 @@ def load_test_resources(test_resources_dir):
...
+def get_current_checkpoint_from_pipeline(
+    pipeline: Pipeline,
+) -> Optional[Checkpoint[GenericCheckpointState]]:
+    azure_ad_source = cast(AzureADSource, pipeline.source)
+    return azure_ad_source.get_current_checkpoint(
+        azure_ad_source.stale_entity_removal_handler.job_id
+    )
+
+
+def stateful_ingest(
```
Comment (hsheth2) [inline metadata-ingestion/tests/integration/azure_ad/test_azure_ad.py:327]: first, this should be named test_stateful_ingest

second, this looks like a lot of copy-pasted code. could you refactor it to reduce the copying?

File: metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py (hunk 1 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py:571]: done

File: metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py (hunk 2 above)
Comment (mohdsiddique) [reply metadata-ingestion/src/datahub/ingestion/source/identity/azure_ad.py:619]: done

File: metadata-ingestion/tests/integration/azure_ad/test_azure_ad.py (hunk 3 above)
Comment (mohdsiddique) [reply metadata-ingestion/tests/integration/azure_ad/test_azure_ad.py]: done

**Pull request merged**